hadoop-tools/hadoop-azure/src/main/java/org/apache/hadoop/fs/azurebfs/services/ListActionTaker.java

```diff
@@ -0,0 +1,178 @@
+      false);
+
+  public ListActionTaker(Path path,
+      AbfsClient abfsClient,
+      int maxConsumptionParallelism,
+      TracingContext tracingContext) {
+    this.path = path;
+    this.abfsClient = (AbfsBlobClient) abfsClient;
+    this.tracingContext = tracingContext;
+    this.maxConsumptionParallelism = maxConsumptionParallelism;
+    executorService = Executors.newFixedThreadPool(maxConsumptionParallelism);
+  }
+
+  abstract boolean takeAction(Path path) throws IOException;
+
+  private boolean takeAction(List<Path> paths) throws IOException {
+    List<Future<Boolean>> futureList = new ArrayList<>();
+    for (Path path : paths) {
+      Future<Boolean> future = executorService.submit(() -> {
+        return takeAction(path);
+      });
+      futureList.add(future);
+    }
+
+    IOException executionException = null;
+    boolean actionResult = true;
+    for (Future<Boolean> future : futureList) {
+      try {
+        Boolean result = future.get();
+        if (!result) {
+          actionResult = false;
+        }
+      } catch (InterruptedException ignored) {
+
+      } catch (ExecutionException e) {
+        executionException = (IOException) e.getCause();
+      }
+    }
+    if (executionException != null) {
+      throw executionException;
+    }
+    return actionResult;
+  }
+
+  /**
+   * Spawns a producer thread that list the children of the path recursively and queue
+   * them in into {@link ListBlobQueue}. On the main thread, it dequeues the
+   * path and supply them to parallel thread for relevant action which is defined
+   * in {@link #takeAction(Path)}.
+   */
+  public boolean listRecursiveAndTakeAction() throws IOException {
+    AbfsConfiguration configuration = abfsClient.getAbfsConfiguration();
+    Thread producerThread = null;
+    try {
+      ListBlobQueue listBlobQueue = new ListBlobQueue(
+          configuration.getProducerQueueMaxSize(), maxConsumptionParallelism);
+      producerThread = new Thread(() -> {
+        try {
+          produceConsumableList(listBlobQueue);
+        } catch (AzureBlobFileSystemException e) {
+          listBlobQueue.markProducerFailure(e);
+        }
+      });
+      producerThread.start();
+
+      while (!listBlobQueue.getIsCompleted()) {
+        List<Path> paths = listBlobQueue.consume();
+        if (paths == null) {
+          continue;
+        }
+        try {
+          boolean resultOnPartAction = takeAction(paths);
+          if (!resultOnPartAction) {
+            return false;
+          }
+        } catch (IOException parallelConsumptionException) {
+          listBlobQueue.markConsumptionFailed();
+          throw parallelConsumptionException;
+        }
+      }
+      return true;
+    } finally {
+      if (producerThread != null) {
+        producerThreadToBeStopped.set(true);
+      }
+      executorService.shutdown();
+    }
+  }
+
+  private void produceConsumableList(final ListBlobQueue listBlobQueue)
```

Comment:
General query here : should we continue to produce paths even after a consumption failure has occurred ?

Reply:
good question, yes. production stop once consumption has failed.

The condition of while loop enforce that

```
while (!producerThreadToBeStopped.get() && continuationToken != null
        && !listBlobQueue.getConsumptionFailed());
```